rebase.go

```diff
@@ -69,6 +72,61 @@ func newRebaseOperationFromC(c *C.git_rebase_operation) *RebaseOperation {
 	return operation
 }
 
+//export commitSignCallback
+func commitSignCallback(_signature *C.git_buf, _signature_field *C.git_buf, _commit_content *C.char, _payload unsafe.Pointer) C.int {
+	opts, ok := pointerHandles.Get(_payload).(*RebaseOptions)
+	if !ok {
+		panic("invalid sign payload")
+	}
+
+	if opts.SigningCallback == nil {
+		return C.GIT_PASSTHROUGH
+	}
+
+	commitContent := C.GoString(_commit_content)
+
+	signature, signatureField, err := opts.SigningCallback(commitContent)
+	if err != nil {
+		if gitError, ok := err.(*GitError); ok {
+			return C.int(gitError.Code)
+		}
+		return C.int(-1)
+	}
+
+	fillBuf := func(bufData string, buf *C.git_buf) error {
+		clen := C.size_t(len(bufData))
+		cstr := unsafe.Pointer(C.CString(bufData))
+		defer C.free(cstr)
+
+		// over-assign by a byte (see below)
+		if int(C.git_buf_grow(buf, clen+1)) != 0 {
+			return errors.New("could not grow buffer")
+		}
+
+		C._go_git_buf_fill_null(buf)
+
+		if int(C.git_buf_set(buf, cstr, clen+1)) != 0 {
+			return errors.New("could not set buffer")
+		}
```

Comment:
OK, second try:

```suggestion
		C.git_buf_clear(buf)
		if int(C.git_buf_put(buf, cstr, clen)) != 0 {
			return errors.New("could not set buffer")
		}
```

[`git_buf_clear()`](https://github.com/libgit2/libgit2/blob/c71321a099373753c22055921c8f538afed5ebb6/src/buffer.c#L152-L163) will remove whatever is in the buffer, and [`git_buf_put()`](https://github.com/libgit2/libgit2/blob/c71321a099373753c22055921c8f538afed5ebb6/src/buffer.c#L223-L238) grows the buffer AND adds the NULL byte at the end.

Reply:
It seems like those functions aren't defined in the 'buffer.h' header and are also not specified in the [documentation](https://libgit2.org/libgit2/#HEAD/search/git_buf_), I tried manually adding the `extern ...` but got an undefined reference error. Are they intended only for internal use?

Reply:
argh, yes u_u they are not marked `GIT_EXTERN`.

but why is it not possible to use

```suggestion
		if int(C.git_buf_set(buf, cstr, clen)) != 0 {
			return errors.New("could not set buffer")
		}
```

? [`git_buf_set`](https://github.com/libgit2/libgit2/blob/c71321a099373753c22055921c8f538afed5ebb6/src/buffer.c#L165-L184) resizes the buffer (plus the space for the NULL byte), and adds the trailing NULL byte. That's even independent of whether the original c-string had a NULL pointer or not. (This time around I did test this suggestion locally, so I'm sure this one does definitely work).

Reply:
I tried this before and although it works in CI (see: github actions on that commit), when running locally I get this error
```
___go_test_github_com_libgit2_git2go_v30: /build/libgit2/src/libgit2-1.0.1/src/rebase.c:1007: rebase_commit__create: Assertion `git_buf_contains_nul(&commit_signature)' failed.
```

Reply:
interesting, that means that we need to have at least one CI configuration where assertions are enabled.

okay, what about

```suggestion
		// libgit2 requires the contents of the buffer to be NULL-terminated.
		// C.CString() guarantees that is the case, so we can safely copy
		// the NULL terminator.
		if int(C.git_buf_set(buf, cstr, clen+1)) != 0 {
			return errors.New("could not set buffer")
		}
```

Reply:
Fixed (and removed the old wrapper)